application-collabora-default/src/main/java/com/xwiki/collabora/internal/AttachmentManager.java

```diff
@@ -0,0 +1,104 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package com.xwiki.collabora.internal;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+
+import javax.inject.Inject;
+import javax.inject.Provider;
+import javax.inject.Singleton;
+
+import org.xwiki.component.annotation.Component;
+import org.xwiki.localization.ContextualLocalizationManager;
+import org.xwiki.model.reference.AttachmentReference;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.rest.XWikiRestException;
+
+import com.xpn.xwiki.XWiki;
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.XWikiException;
+import com.xpn.xwiki.doc.XWikiAttachment;
+import com.xpn.xwiki.doc.XWikiDocument;
+
+/**
+ * Manage document attachments.
+ *
+ * @version $Id$
+ * @since 1.0
+ */
+@Component(roles = AttachmentManager.class)
+@Singleton
+public class AttachmentManager
+{
+    @Inject
+    private ContextualLocalizationManager contextualLocalizationManager;
+
+    @Inject
+    private Provider<XWikiContext> contextProvider;
+
+    /**
+     * Create or update attachment with given content.
+     *
+     * @param attachmentReference reference of the attachment
+     * @param content the new attachment content
+     * @return a modified {@link XWikiAttachment}
+     * @throws XWikiRestException If an exception occurs while the Attachment is created or updated.
+     */
+    public XWikiAttachment createOrUpdateAttachment(AttachmentReference attachmentReference, byte[] content)
+        throws XWikiRestException
+    {
+        XWikiContext xcontext = this.contextProvider.get();
+        XWiki xwiki = xcontext.getWiki();
+        DocumentReference documentReference = attachmentReference.getDocumentReference();
+
+        try {
+            // We clone the document because we're going to modify it and we shouldn't modify the cached instance.
+            XWikiDocument document = xwiki.getDocument(documentReference, xcontext).clone();
+            XWikiAttachment attachment =
+                document.setAttachment(attachmentReference.getName(), new ByteArrayInputStream(content), xcontext);
+            attachment.setAuthorReference(xcontext.getUserReference());
+
+            document.setAuthorReference(xcontext.getUserReference());
+            xwiki.saveDocument(document,
+                this.contextualLocalizationManager.getTranslationPlain("collabora.save.comment"), xcontext);
+
+            return attachment;
+        } catch (XWikiException | IOException e) {
+            throw new XWikiRestException(
```

Comment:
It feels strange to throw a REST exception from a component outside of the ``*.rest.*`` package. REST is not the only way in which you can expose / publish your API. You can also have a script service, in which case it would be strange to get a REST exception logged from a script service. Even if REST is the only way in which you plan to expose your API, it's still cleaner to separate the two (model / core API vs. REST API).